pay/contracts/subscription/HectorSubscription.sol

```diff
@@ -0,0 +1,319 @@
+
+        Plan memory plan = plans[_planId];
+
+        if (balanceOf[msg.sender][plan.token] < plan.amount)
+            revert PAYER_IN_DEBT();
+
+        balanceOf[msg.sender][plan.token] -= plan.amount;
+        IERC20(plan.token).safeTransfer(treasury, plan.amount);
+
+        subscriptions[msg.sender] = Subscription({
+            planId: _planId,
+            expiredAt: uint48(block.timestamp + plan.period)
+        });
+
+        emit SubscriptionCreated(msg.sender, _planId);
+    }
+
+    function depositAndCreateSubscription(uint256 _planId, uint256 _amount)
+        external
+    {
+        deposit(plans[_planId].token, _amount);
+        createSubscription(_planId);
+    }
+
+    function syncSubscription(address from) public {
+        Subscription storage subscription = subscriptions[from];
+        uint256 planId = subscription.planId;
+
+        if (planId == 0) revert INACTIVE_SUBSCRIPTION();
+
+        // before expiration
+        if (block.timestamp < subscription.expiredAt) {}
+        // after expiration
+        else {
+            Plan memory plan = plans[planId];
+            uint256 count = (block.timestamp -
+                subscription.expiredAt +
+                plan.period) / plan.period;
+            uint256 amount = plan.amount * count;
+            uint256 balance = balanceOf[from][plan.token];
+
+            // not active for now
+            if (balance < amount) {
+                count = balance / plan.amount;
+                amount = plan.amount * count;
+            }
+
+            if (count > 0) {
+                balanceOf[from][plan.token] -= amount;
+                subscription.expiredAt += uint48(plan.period * count);
+
+                IERC20(plan.token).transfer(treasury, amount);
+
+                emit SubscriptionSynced(from, subscription.expiredAt);
+            }
+        }
+    }
+
+    function cancelSubscription() external {
+        syncSubscription(msg.sender);
+
+        subscriptions[msg.sender].planId = 0;
+
+        emit SubscriptionCancelled(msg.sender);
+    }
+
+    function modifySubscription(uint256 _newPlanId)
+        external
+        onlyValidPlan(_newPlanId)
+    {
+        syncSubscription(msg.sender);
+
+        subscriptions[msg.sender].planId = _newPlanId;
+
+        emit SubscriptionModified(msg.sender, _newPlanId);
+    }
+
+    function withdraw(address _token, uint256 _amount) external {
+        if (subscriptions[msg.sender].planId > 0) {
+            syncSubscription(msg.sender);
+        }
+
+        if (_token == address(0)) revert INVALID_ADDRESS();
+        if (_amount == 0) revert INVALID_AMOUNT();
+        if (balanceOf[msg.sender][_token] < _amount) revert INVALID_AMOUNT();
+
+        balanceOf[msg.sender][_token] -= _amount;
+
+        IERC20(_token).safeTransfer(msg.sender, _amount);
+
+        emit PayerWithdraw(msg.sender, _token, _amount);
+    }
+
+    function withdrawAll(address _token) external returns (uint256 amount) {
```

Comment:
@bejaxer we need an admin function to facilitate the refunds for user. 
- Create an array of moderators
- Wallet in the list of moderators can execute this func
- Input parameters: amount, recipient
- Make sure recipient has an existing subscription; otherwise don't allow refund
- Emit events
- Visible in graphql

Reply:
@bejaxer we also need an admin function to withdraw any existing fund in this contract and transfer to treasury wallet.
- Create an array of moderators
- Wallet in the list of moderators can execute this func
- Input parameters: none
- Emit events